activation/activation.go

```diff
@@ -135,133 +139,69 @@ func (b *Builder) Start() {
 
 // Stop stops the atx builder.
 func (b *Builder) Stop() {
-	b.finished <- 0
 	close(b.stop)
 }
 
-func (b Builder) SignAtx(atx *types.ActivationTx) (*types.ActivationTx, error) {
-	bts, err := types.InterfaceToBytes(atx.InnerActivationTx)
-	if err != nil {
-		return nil, err
-	}
-	atx.Sig = b.Sign(bts)
-	return atx, nil
+func (b Builder) SignAtx(atx *types.ActivationTx) error {
+	return SignAtx(b, atx)
 }
 
+type StopRequestedError struct{}
+
+func (s StopRequestedError) Error() string { return "stop requested" }
+
 // loop is the main loop that tries to create an atx per tick received from the global clock
 func (b *Builder) loop() {
 	err := b.loadChallenge()
 	if err != nil {
 		log.Info("challenge not loaded: %s", err)
 	}
+	select {
+	case <-b.stop:
+		return
+	case <-b.initDone:
+	}
+	<-b.layerClock.AwaitLayer(1) // this doesn't block unless layer 1 hasn't arrived yet
 	for {
 		select {
 		case <-b.stop:
 			return
-		case layer := <-b.timer:
-			if !b.isSynced() {
-				b.log.Info("cannot create atx : not synced")
-				break
-			}
-
-			if atomic.LoadInt32(&b.initStatus) != InitDone {
-				b.log.Info("post is not initialized yet, not building Nipst")
-				break
-			}
-			if b.working {
-				break
-			}
-			b.working = true
-			go func() {
-				epoch := layer.GetEpoch(b.layersPerEpoch)
-				err := b.PublishActivationTx(epoch)
-				if err != nil {
-					if err == tooSoonErr {
-						b.log.Warning("cannot create atx in epoch %v: %v", epoch, err)
-					} else {
-						b.log.Error("cannot create atx in epoch %v: %v", epoch, err)
-					}
-					events.Publish(events.AtxCreated{false, "", uint64(epoch)})
-				}
-
-				b.finished <- layer
-			}()
-		case <-b.finished:
-			b.working = false
+		default:
 		}
-	}
-}
-
-type alreadyPublishedErr struct{}
-
-func (alreadyPublishedErr) Error() string { return "already published" }
-
-func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
-	if b.prevATX == nil {
-		prevAtxId, err := b.GetPrevAtxId(b.nodeId)
-		if err != nil {
-			b.log.Info("no prev ATX found, starting fresh")
-		} else {
-			b.prevATX, err = b.db.GetAtxHeader(prevAtxId)
-			if err != nil {
-				// TODO: handle inconsistent state
-				b.log.Panic("prevAtx (id: %v) not found in DB -- inconsistent state", prevAtxId.ShortString())
+		if err := b.PublishActivationTx(); err != nil {
+			if _, stopRequested := err.(StopRequestedError); stopRequested {
+				return
 			}
+			b.log.With().Error("failed to publish ATX", log.Err(err))
+			events.Publish(events.AtxCreated{false, "", uint64(b.currentEpoch())})
+			<-b.layerClock.AwaitLayer(b.layerClock.GetCurrentLayer() + 1)
```

Comment:
why do we need to wait an entire layer to resend?

Reply:
This was a lazy choice. This is what the previous implementation did.

I know that one class of errors, specifically PoET errors, will require special handling. There are three kinds of PoET errors: 
1. No response to registration requests - if the node is aware of another PoET _(no implemented)_ it should try another one.
2. PoET membership proof doesn't include our registration and we have a signed receipt _(not implemented)_ - blacklist PoET _(not implemented)_ and publish fraud proof _(not implemented)_.
3. Waiting for PoET proof timed out - blacklist PoET _(not implemented)_. Proving this happened is non-trivial, perhaps we can piggyback the hare, but not sure it's worth it.

As you can see none of this is possible right now (created issue for this just now #1673).

We could map out the other errors and see how we should handle each one, maybe the best handling for most of them really is to just retry immediately. I'll look at it now.